src/driver-marsgaming/marsgaming-commit.c

```diff
@@ -0,0 +1,144 @@
+	if (button_data.is_present)
+		*button_info = button_data.button_info;
+	return 0;
+}
+
+static int
+marsgaming_commit_led(struct ratbag_led *led)
+{
+	if (!led->dirty)
+		return 0;
+	marsgaming_command_profile_set_led(led);
+	return 0;
+}
+
+static int
+marsgaming_commit_profile_report_rate(struct ratbag_profile *profile)
+{
+	if (!profile->rate_dirty)
+		return 0;
+
+	uint8_t polling_interval = 1000 / ratbag_profile_get_report_rate(profile);
+	marsgaming_command_profile_set_polling_interval(profile, polling_interval);
+	return 0;
+}
+
+static int
+marsgaming_commit_profile_buttons(struct ratbag_profile *profile)
+{
+	bool buttons_dirty = false;
+	struct ratbag_button *button;
+	ratbag_profile_for_each_button(profile, button) {
+		if (button->dirty) {
+			buttons_dirty = true;
+			break;
+		}
+	}
+	if (!buttons_dirty)
+		return 0;
+	ratbag_profile_for_each_button(profile, button) {
+		if (!button->dirty)
+			continue;
+		marsgaming_commit_button(button);
+	}
+	marsgaming_command_profile_set_buttons(profile);
+	return 0;
+}
+
+static int
+marsgaming_commit_profile_leds(struct ratbag_profile *profile)
+{
+	struct ratbag_led *led;
+	ratbag_profile_for_each_led(profile, led) {
+		marsgaming_commit_led(led);
+	}
+	return 0;
+}
+
+static int
+marsgaming_commit_profile_resolutions(struct ratbag_profile *profile)
+{
+	bool resolutions_dirty = false;
+	struct ratbag_resolution *resolution;
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (resolution->dirty) {
+			resolutions_dirty = true;
+			break;
+		}
+	}
+	if (!resolutions_dirty)
+		return 0;
+	struct marsgaming_profile_drv_data *profile_data = marsgaming_profile_get_drv_data(profile);
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+		// Modify the drv_data report so we can send it to the mouse
+		struct marsgaming_report_resolution_info *resolution_info = &profile_data->resolutions_report.resolutions[resolution->index];
+		resolution_info->enabled = true;
+		resolution_info->x_res = resolution->dpi_x / MARSGAMING_MM4_RES_SCALING;
+		resolution_info->y_res = resolution->dpi_y / MARSGAMING_MM4_RES_SCALING;
+		resolution_info->led_bitset = ~((~0x0U) << resolution->index);
+	}
+
+	marsgaming_command_profile_set_resolutions(profile);
+
+	return 0;
+}
+
+static int
+marsgaming_commit_profile(struct ratbag_profile *profile)
+{
+	if (!profile->dirty)
+		return 0;
+	marsgaming_commit_profile_report_rate(profile);
+	marsgaming_commit_profile_resolutions(profile);
+	marsgaming_commit_profile_buttons(profile);
+	marsgaming_commit_profile_leds(profile);
+	return 0;
+}
+
+static int
+marsgaming_commit_profiles(struct ratbag_device *device)
+{
+	uint8_t current_profile = marsgaming_query_current_profile(device);
+	struct ratbag_profile *profile;
+	ratbag_device_for_each_profile(device, profile) {
+		profile->is_active = current_profile == profile->index; // TODO: Check if we should be modifying this
```

Comment:
Looks like this will simply mark every single profile as active, which is incorrect, so probably this is not needed.

Reply:
I don't think that's the case, no? Only the profile whose index matches `current_profile` will be set as active, right?

The TODO was not about this question though. The issue comes when the user modifies the active profile after probe but before commit. Only the led of the active profile should be changed in `marsgaming-command.c:43`, but I don't know if modifying this here can cause any issues.

Reply:
> Looks like this will simply mark every single profile as active, which is incorrect, so probably this is not needed.

I was wrong here.

---

> The TODO was not about this question though

Could you put this explanation to the TODO itself? Otherwise it's not really clear what it means.

Reply:
I have expanded on the comment to make the intent clearer. The doubt still remains, although during development I did not see any unexpected behavior.